app/src/main/java/com/asfoundation/wallet/topup/TopUpFragmentPresenter.kt

```diff
@@ -158,4 +186,116 @@ class TopUpFragmentPresenter(private val view: TopUpFragmentView,
           }
         }
   }
+
+  private fun handleInvalidValueInput(packageName: String, topUpData: TopUpData) {
+    if (topUpData.currency.fiatValue != DEFAULT_VALUE) {
+      val value =
+          FiatValue(BigDecimal(topUpData.currency.fiatValue), topUpData.currency.fiatCurrencyCode,
+              topUpData.currency.fiatCurrencySymbol)
+      disposables.add(fetchMinAndMaxValues()
+          .subscribeOn(networkScheduler)
+          .observeOn(viewScheduler)
+          .doOnSuccess {
+            showValueWarning(it.first, it.second, value)
+            handleShowBonus(packageName, topUpData, it.first, it.second, value)
+          }
+          .subscribe())
+    }
+  }
+
+  private fun fetchMinAndMaxValues(): Single<Pair<FiatValue, FiatValue>> {
+    return Single.zip(interactor.getMaxTopUpValue(), interactor.getMinTopUpValue(),
+        BiFunction { maxValue: FiatValue, minValue: FiatValue -> Pair(maxValue, minValue) })
+  }
+
+  private fun handleShowBonus(appPackage: String, topUpData: TopUpData, maxValue: FiatValue,
+                              minValue: FiatValue, value: FiatValue) {
+    if (value.amount < minValue.amount || value.amount > maxValue.amount) {
+      view.hideBonus()
+      view.changeMainValueColor(false)
+    } else {
+      view.changeMainValueColor(true)
+      disposables.add(loadBonusIntoView(appPackage, topUpData.currency.fiatValue,
+          topUpData.currency.fiatCurrencyCode)
+          .doOnNext { view.setNextButtonState(hasValidData(topUpData)) }
+          .subscribe())
+    }
+  }
+
+  private fun handleDefaultValueChips() {
+    disposables.add(interactor.getDefaultValues()
+        .subscribeOn(networkScheduler)
+        .observeOn(viewScheduler)
+        .map { view.setupDefaultValueChips(it) }
+        .subscribe())
+  }
+
+  private fun handleValueChipsClick() {
+    disposables.add(view.getChipsClick()
+        .doOnNext { index ->
```

Comment:
Add a observeOn, just to absolutely guarantee that this will not run on the io thread (as it is right now it should work, but there’s a small probability that it can’t)